haddock-api/src/Haddock/Backends/Hyperlinker/Ast.hs

```diff
@@ -0,0 +1,192 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Haddock.Backends.Hyperlinker.Ast
+    ( enrich
+    , RichToken(..), TokenDetails(..), rtkName
+    ) where
+
+import Haddock.Backends.Hyperlinker.Parser
+
+import qualified GHC
+
+import Control.Applicative
+import Data.Data
+import Data.Maybe
+
+data RichToken = RichToken
+    { rtkToken :: Token
+    , rtkDetails :: Maybe TokenDetails
+    }
+
+data TokenDetails
+    = RtkVar GHC.Name
+    | RtkType GHC.Name
+    | RtkBind GHC.Name
+    | RtkDecl GHC.Name
+    | RtkModule GHC.ModuleName
+
+rtkName :: TokenDetails -> Either GHC.Name GHC.ModuleName
+rtkName (RtkVar name) = Left name
+rtkName (RtkType name) = Left name
+rtkName (RtkBind name) = Left name
+rtkName (RtkDecl name) = Left name
+rtkName (RtkModule name) = Right name
+
+-- | Add more detailed information to token stream using GHC API.
+enrich :: GHC.RenamedSource -> [Token] -> [RichToken]
+enrich src =
+    map $ \token -> RichToken
+        { rtkToken = token
+        , rtkDetails = enrichToken token detailsMap
+        }
+  where
+    detailsMap = concatMap ($ src)
+        [ variables
+        , types
+        , decls
+        , binds
+        , imports
+        ]
+
+-- | A map containing association between source locations and "details" of
+-- this location.
+--
+-- For the time being, it is just a list of pairs. However, looking up things
+-- in such structure has linear complexity. We cannot use any hashmap-like
+-- stuff because source locations are not ordered. In the future, this should
+-- be replaced with interval tree data structure.
+type DetailsMap = [(GHC.SrcSpan, TokenDetails)]
```

Comment:
When I wrote that they are not ordered I meant that source spans do not form any sane total order (spans are intervals and intervals are just sets, and sets are only partially ordered). They might be ordered in a Haskell sense (by using ordinary lexical ordering on span coordinates for example), but it is not very helpful as I am looking things up by checking for interval intersection, hence the `Map` has no use here. Suitable data structure would be either [segment tree](https://en.wikipedia.org/wiki/Segment_tree) or [interval tree](https://en.wikipedia.org/wiki/Interval_tree) but these are not available in core as far as I know.